keps/sig-node/20190517-image-decryption.md

```diff
@@ -0,0 +1,345 @@
+  </snip>
+}
+```
+
+```go
+type ServiceAccount struct {
+  <snip>
+  // ImageDecryptSecrets is a list of references to secrets in the same namespace to use for decrypting any encrypted images
+  // in pods that reference this ServiceAccount. 
+  // +optional
+  ImageDecryptSecrets []LocalObjectReference `json:"imageDecryptSecrets,omitempty" protobuf:"bytes,5,rep,name=imageDecryptSecrets"`
+  </snip>
+```
+
+An unspecified `nil` or empty `""` ImageDecryptSecret is equivalent to the backwards-compatible
+default behavior as if the ImageDecryptSecret feature is disabled.
+
+#### Examples
+
+Suppose we operate a cluster that lets users create a secret of type `ImageDecryptSecret`. 
+
+For the kubectl command line, we propose that the user create a secret of type `image-decrypt` and give it a name. One or multiple private keys may then be stored under this secret.
+
+```bash
+kubectl create secret image-decrypt <secret name> --decrypt-secret=/path/to/private_key[:<password>] [--decrypt-secret=/path/to/private_key[:<password>]]
+```
+`private key` - Private keys represented in format DER/PEM or GPG private keyrings.  These keys will be used by JWE/PKCS7/PGP protocols.
+
+`password` (optional) - password in *cleartext* if the given private key is protected with it.
+
+For example,
+```bash
+kubectl create secret image-decrypt keysecret --decrypt-secret="/home/user/keys/private.key":"password"
+```
+
+When a user creates a workload, they can choose the desired ImageDecryptSecret to use 
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: nginx
+  labels:
+    name: nginx
+spec:
+  containers:
+  - name: nginx
+    image: localhost:5000/nginx:enc
+    ports:
+    - containerPort: 80
+  imageDecryptSecrets:
+  - name: keysecret
+```
+
+ImageDecryptSecrets can be added to the `service account` by,
+
+```bash
+kubectl patch serviceaccount <account name> -p '{"imageDecryptSecrets":[{"name":<secret name>]}'
+```
+or while creating a secret account,
+```yaml
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  creationTimestamp: 2015-08-07T22:02:39Z
+  name: default
+  namespace: default
+  selfLink: /api/v1/namespaces/default/serviceaccounts/default
+  uid: 052fb0f4-3d50-11e5-b066-42010af0d7b6
+secrets:
+- name: default-token-uudge
+imagePullSecrets:
+- name: myregistrykey
+imageDecryptSecrets:
+- name: <secret name>
+```
+
+For example,
+```bash
+kubectl patch serviceaccount default -p '{"imageDecryptSecrets":[{"name":"keysecret"}]}'
+```
+
+#### Key Secret Definition
+
+The privake keys are extracted from `ImageDecryptSecret` and passed to the CRI by bundling them in `DecryptParam` as a part of `PullImageRequest` (see, [Image Handler](#Image-handler)) and `CreateContainerRequest` (see, [Container Handler](#Container-handler)):
+
+
+```protobuf
+// DecryptParam represents a single private key (and optional password) that can be sent to the CRI via `PullImageRequest` as well as `CreateContainerRequest`
+message DecryptParam {
```

Comment:
Would be good to make the name more specific. Maybe `ImageDecryptParam`?

Reply:
Yes, It makes sense.